### PR TITLE
Block editor: add a keyboard shortcut to create group from the selected blocks

### LINKED ISSUE #42227: Add a keyboard shortcut to create group
A common pattern in design software is that selected frames can be converted to a group via <kbd>CMD</kbd>+<kbd>G</kbd> (or similar) keyboard shortcut.

Now that site editing (where grouping is common) is becoming more prevalent it's worth considering adding a shortcut to do perform this action.

### PR DESCRIPTION
**Pull request by t-hamano**
Closes #42227

## What?
This PR adds a keyboard shortcut to create a group block from the selected blocks.

## Why?
Many design tools provide shortcuts for grouping elements. It would make sense to add a similar shortcut, especially since the grouping of blocks should be done frequently in the site editing.

## How?
To explore which key combinations were available, I tested all variations that included the letter `G`. The list can be found in [this comment](https://github.com/WordPress/gutenberg/issues/42227#issuecomment-1374842421).

The results show that the following two modifiers are available:

- `primaryAlt`: `⌘ command` + `⌥ option` + `G` on Mac / `ctrl` + `alt` + `G` on Window
- `secondary`: `⌘ command` + `shift` + `⌥ option` + `G` on Mac / `ctrl` + `shift` + `alt` + `G`

~I chose `primaryAlt` because the `secondary` modifier could be used for ungrouping (unwrapping) in the future.~

~However, as noted in this comment, it appears that Ctrl + Alt is usually used for Windows OS-level functions. I would appreciate your advice on whether it makes sense to add this shortcut and whether it will cause any confusion to the user.~

Update: After some discussion, we adopted the simpler `⌘ command + G` (`ctrl + G`) as a shortcut.

Also, this shortcut will not allow to create a group block if a single block is selected. This is consistent with the behavior of group block transformation.

## Testing Instructions

- Insert multiple blocks and select them all.
- On Windows, press `ctrl` + `alt` + `G`. On Mac, press `⌘ command` + `⌥ option` + `G`.
- Confirm that the selected block is wrapped in a group block.
- The second shortcut should not be executed because the group block is selected.

## Screenshots or screencast

https://user-images.githubusercontent.com/54422211/211314612-b1399643-c770-4774-a323-ebb95e0c2fc4.mp4

### CODE REVIEW

**Comment by jameskoster** [comment]
@t-hamano any reason this never got added? 🤔

**Comment by t-hamano** [comment]
I think this PR makes sense, but maybe it got buried because there are so many open PRs out there 😅 There is a conflict going on and I would like to resolve this and ~~also add an E2E test~~(Already added!).

**Comment by jameskoster** [comment]
> this shortcut will not allow to create a group block if a single block is selected

I'm not sure about this part. You might want to wrap a block in a container to add extra styles, or layout.

> The second shortcut should not be executed because the group block is selected.

This one I'm 50:50 on. I understand the reasoning, but am a little wary it will make legitimate use cases harder to execute.

**Comment by t-hamano** [comment]
> this shortcut will not allow to create a group block if a single block is selected

I don't have a strong opinion on this, but it also serves to prevent multi-level group blocks from being created when shortcuts are executed consecutively. What do you think?

https://github.com/WordPress/gutenberg/assets/54422211/23ca7cf5-2544-4e88-9cda-f3d8d4129271

**Comment by jasmussen** [comment]
Thanks for the PR, sorry for missing this and appreciate the ping. I'm a big fan of adding grouping and ungrouping shortcuts, let's also add them to the ellipsis menu as tips, I notice they're missing there now even if they are in the keyboard shortcuts.

Another aspect, for shortcuts, we should probably reserve Shift for ungrouping. I.e. `[main modifier(s)] + g` for grouping, `[main modifier(s)] + Shift + g` for ungrouping. I realize this only adds for grouping, but important to consider a future ungrouping.

For me, option + command + g opens Google Drive (global app I have running in the background on my Mac). That's a damn bummer. Not necessarily a blocker, but a bummer. Outside of the plain Cmd + g and Cmd + Shift + g shortcuts, for my part that leaves Ctrl + Cmd + g and Ctrl + Cmd + Shift + g. Would those work?

That aside, and I know this was discussed in the other thread, is there any reason why the plain Cmd + g and Cmd + Shift + g shortcuts can't work? I understand that they are used by the browser for "find next", but can we do the same as we do for the command palette and make the shortcut contextual? I.e. if you have two or more blocks selected, Cmd + g groups. If you don't have any blocks selected, you "find next". For example in Figma, these shortcuts work: they simply override the browser shortcuts.

**Comment by t-hamano** [comment]
I changed the key combination as follows.

- Windows:  `ctrl + G`
- Mac: `⌘ command + G`

This should match the command for grouping in Figma on both Windows and Mac, is this correct? Please let me know if this change makes sense since I can't test it on a Mac 😅

However, this key combination should override the browser's search shortcut on Windows and Mac. So I think this shortcut should only be allowed in the context of multiple blocks being selected.

> Another aspect, for shortcuts, we should probably reserve Shift for ungrouping. I.e. `[main modifier(s)] + g` for grouping, `[main modifier(s)] + Shift + g` for ungrouping. I realize this only adds for grouping, but important to consider a future ungrouping.

Yes, this could be considered in follow-up. Or maybe we can include it in this PR.

If we set the shortcut for grouping as described here, the shortcut for ungrouping would be:

- Windows:  `ctrl + shift + G`
- Mac: `⌘ command + Shift + G`

**Comment by t-hamano** [comment]
> let's also add them to the ellipsis menu as tips, I notice they're missing there now even if they are in the keyboard shortcuts.

I made the shortcut key appear only when multiple blocks are selected. Because this current PR does not allow grouping when a single block is selected.

A single block is selected:


image

Multiple blocks are selected:

image

**Review by jasmussen — APPROVED** [review]
Behavior wise this is solid. There's a shortcut indication in the menu:

shortcut

⌘G works great for grouping when you have two blocks selected. When you do not have blocks selected, ⌘G invokes the browser shortcut for find next. This works great.

There's _one_ followup to do, I noticed that when you have a group or other container block selected, the shortcut for the Group does not show up:

Screenshot 2024-05-02 at 08 35 24

But this is an excellent quality of life feature, very nice work. I can't wait for a followup that adds ⌘⇧G for ungrouping 👌

File: packages/block-editor/src/components/block-tools/index.js
Comment (talldan) [inline packages/block-editor/src/components/block-tools/index.js]: A spoken message when grouping is successful might be a good affordance. I tested with voiceover and there's not any indication that anything happens.

**Review by talldan — APPROVED** [review]
Looks great - one suggestion, but I think you can address it and then merge, or get feedback from a more experienced a11y contributor first if you want to.

File: packages/block-editor/src/components/block-tools/index.js
Comment (t-hamano) [reply packages/block-editor/src/components/block-tools/index.js]: Updated by 8cf7f34

When you run the shortcut, you should hear "Selected blocks are grouped.".

Video tested with NVDA:


https://github.com/WordPress/gutenberg/assets/54422211/7927b1b0-fd5a-4353-86a9-6d5588030270

**Comment by t-hamano** [comment]
> but I think you can address it and then merge, or get feedback from a more experienced a11y contributor first if you want to.

Let's get feedback before merging this PR. @afercia, @alexstine, if you have any feedback regarding this PR, please let us know.

### Summary

- This PR adds a shortcut to wrap selected blocks with a Group block.
- At first, I considered using `⌘ command + ⌥ option + G` (`ctrl + alt + G`), but after some discussion we decided to use `⌘ command + G` (`ctrl + G`).
- This shortcut conflicts with the browser's search shortcut, so it only performs grouping when multiple blocks are selected.

File: packages/block-editor/src/components/block-tools/index.js
Comment (t-hamano) [reply packages/block-editor/src/components/block-tools/index.js]: By the way, we may need to add a spoken message to other shortcuts as well. See #61168.

**Comment by bacoords** [comment]
A few notes from my testing (MacOS)
- The keyboard shortcut shows up next to "Group" in the dropdown when you have multiple blocks selected but not when you have only one block selected, even though the option to "Group" still exists. This could be related to another issue with the fact that grouping/transforming have sort of a blurred lines approach and can work in different ways for unclear reasons:  (#58042)
- The shortcut does not seem to work for me when selecting items in the list view, only when selecting blocks on the canvas.

I would also second the idea that if we're adding a shortcut for Group it makes sense to do Ungroup as well (with the `shift` modifier).

But overall great work and this is going to be a very popular enhancement.

**Comment by jameskoster** [comment]
> but it also serves to prevent multi-level group blocks from being created when shortcuts are executed consecutively.

To be honest I don't feel this guardrail is necessary. In most design software, repetition of the shortcut creates additional groups. In not allowing this, I worry it might be frustrating in situations you _want_ nested groups, rare as it may be. Ultimately it feels like something we could add later based on feedback.

**Comment by t-hamano** [comment]
>The shortcut does not seem to work for me when selecting items in the list view, only when selecting blocks on the canvas.

Nice catch! Added shortcut to work in the list view as well.

> > but it also serves to prevent multi-level group blocks from being created when shortcuts are executed consecutively.
> 
> To be honest I don't feel this guardrail is necessary.

If we remove this guardrail, in most cases you will override the browser's search shortcut (Find Next / Find Previous). If a user wants to use the browser's search shortcut, they are forced to move focus away from the editor canvas or list view.

This is my only concern, but I'd love to hear what the accessibility team has to say.

**Comment by jasmussen** [comment]
It could be reasonable to start with the current PR behavior, and then remove the guardrail as a separate discussion at a later time, if the grouping behavior feels unintuitive.

**Comment by alexstine** [comment]
@t-hamano Thanks for waiting on feedback, much appreciated. Hoping I can start reviewing faster soon.

Do not have time to do a code review at this point but the accessibility is good. The announcement is working as expected.

Let's not remove the guardrail in this PR because we need to figure out how to communicate to non-sighted users that multiple nested groups could be created depending on how many times they use the shortcut. Honestly, I would have a hard time understanding why someone would want a group inside a group.

The only issue I could find is when grouping blocks from the list view, the editor canvas steals focus. I think this has come up in other PRs and if you don't have an easy way to fix it, let's figure out how to solve the actual problem in a follow-up.

Thanks.

File: packages/block-editor/src/components/list-view/block.js
Comment (andrewserong) [inline packages/block-editor/src/components/list-view/block.js]: In the List View, rather than using the selected block client ids, I think it's generally better to use `const { blocksToUpdate } = getBlocksToUpdate();` as used in the insert before and after keyboard shortcuts above. This will ensure that if the user is focused on a block that isn't part of the selection, then the action is performed on the currently focused block. If the focused block is part of the selection, then the whole selection is used.

File: packages/block-editor/src/components/list-view/block.js
Comment (andrewserong) [inline packages/block-editor/src/components/list-view/block.js]: To fix the issue of the canvas stealing focus that Alex mentioned, we might be able to use a similar approach as in the insert before and after keyboard shortcuts above, where we grab the selected blocks after calling `replaceBlocks` and then call the following to shift focus back to the list view again:

```
			const newlySelectedBlocks = getSelectedBlockClientIds();
			// Focus the first block of the newly inserted blocks, to keep focus within the list view.
			updateFocusAndSelection( newlySelectedBlocks[ 0 ], false );
```

**Review by andrewserong — COMMENTED** [review]
This is such a nice enhancement! I've just left a couple of comments about the list view behaviour, which I think will help address the notes that Alex raised regarding the focus shift when using the shortcut in the list view.

File: packages/block-editor/src/components/list-view/block.js
Comment (andrewserong) [reply packages/block-editor/src/components/list-view/block.js]: I just realised this might not be particularly clear as to why it matters. Here's a quick example video of the behaviour as it is currently, where if I move up the list view to outside the selection, the keyboard shortcut acts on the selected blocks instead of the focused one:

https://github.com/WordPress/gutenberg/assets/14988353/f66c745c-e4e4-44ad-b5e6-7786833a1382

**Comment by t-hamano** [comment]
@alexstine @andrewserong Thanks for the review!

> Let's not remove the guardrail in this PR because we need to figure out how to communicate to non-sighted users that multiple nested groups could be created depending on how many times they use the shortcut. Honestly, I would have a hard time understanding why someone would want a group inside a group.

I see, let's keep the current specs.

> In the List View, rather than using the selected block client ids, I think it's generally better to use `const { blocksToUpdate } = getBlocksToUpdate();`
> where we grab the selected blocks after calling `replaceBlocks` and then call the following to shift focus back to the list view again:

I have also addressed this approach. [This part](https://github.com/WordPress/gutenberg/pull/46972/files#diff-c616112dd9cb24e1c02eddc8bd10f0a815705d8dbb707dcc7fded7a010b141aeR1011) of the E2E test should ensure that the focus remains on the list view when you run the shortcut in the list view.

**Comment by jameskoster** [comment]
> we need to figure out how to communicate to non-sighted users that multiple nested groups could be created depending on how many times they use the shortcut

Could a Snackbar work, similar to the 'Copy all blocks' action?

File: packages/block-editor/src/components/list-view/block.js
Comment (andrewserong) [inline packages/block-editor/src/components/list-view/block.js:337]: Why is the check that the length is greater than one? If a single block is selected or focused, should we still be able to group it? In the editor canvas, I think it's good to only allow it in a multi-selection, but it could still be useful to allow it in the list view, otherwise this is what happens when using the shortcut over a focused item that isn't part of the selection:

https://github.com/WordPress/gutenberg/assets/14988353/f73ef906-63d0-45d6-81b5-6a8f651d7b24

The above video demoes grouping two blocks, then moving up to a block outside the selection and pressing the keyboard shortcut.

**Review by andrewserong — COMMENTED** [review]
Thanks for the update, this feels very close to me! Just noticed something when focusing over a single item in the list view, so I've left a comment.

File: packages/block-editor/src/components/list-view/block.js
Comment (t-hamano) [reply packages/block-editor/src/components/list-view/block.js:337]: The reason for checking that the length is greater than 1 is to be consistent with the editor canvas. When I run the shortcut on a single item that has focus and is not in a selection, the browser's search window opens. I think this is the expected behavior for now, but is the same behavior happening in your environment?

browser-search

Comment (andrewserong) [reply packages/block-editor/src/components/list-view/block.js:337]: Yes, that's what's happening for me, too. All good to leave it as-is for now, and we can always revisit in follow-ups if folks run into any issues. The main value is to group multiple items together so I doubt that in practice anyone will really struggle with the current behaviour.

**Review by andrewserong — APPROVED** [review]
Thanks for all the back and forth and discussion! I think this is in a good place to land, and we can always continue to iterate based on feedback.

LGTM! 🚀

**Comment by t-hamano** [comment]
Thank you everyone for your reviews! I think everything is ready, so let's merge this PR. If we have any accessibility issues or spec concerns, I will be happy to follow up and address them.

**Pull request merged**
